circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/CollectRetained.kt

```diff
@@ -40,3 +40,32 @@ public fun <T : R, R> Flow<T>.collectAsRetainedState(
       collect { value = it }
     } else withContext(context) { collect { value = it } }
   }
+
+/**
+ * Produces and collects values from a [Flow] provided by the [producer] lambda and represents its
+ * latest value via retained [State]. Every time there would be a new value posted into the [Flow],
+ * the returned [State] will be updated causing recomposition of every [State.value] usage.
+ *
+ * This function safely produces and collects a Flow by using a single retained state with unified
+ * inputs. The Flow is produced within the retained state's producer block, ensuring proper scoping
+ * and preventing memory leaks from retaining Flow instances across configuration changes.
+ *
+ * @param inputs A set of inputs such that, when any of them have changed, will cause the state to
+ *   reset and [producer] to be rerun
+ * @param initial the initial value for the state
+ * @param context [CoroutineContext] to use for collecting.
+ * @param producer a suspend lambda that provides the [Flow] to collect from
+ */
+@Composable
+public fun <T : R, R> produceAndCollectAsRetainedState(
+  vararg inputs: Any?,
+  initial: R,
+  context: CoroutineContext = EmptyCoroutineContext,
+  producer: suspend () -> Flow<T>,
+): State<R> =
+  produceRetainedState(initial, *inputs) {
```

Comment:
Still think we should include the producer in the keys for correctness, even if it's unnecessary most of the time